scripts/changelog_check.rb

```diff
@@ -150,13 +151,14 @@ def generate_changelog(git_log)
 # Entries with the same category and change are grouped into one changelog line so that we can
 # support multi-PR changes.
 def format_changelog(changelog_entries)
-  changelog_entries = changelog_entries.group_by { |entry| [entry.category, entry.change] }
+  changelog_entries = changelog_entries.
+    sort_by(&:subcategory).
+    group_by { |entry| [entry.category, entry.change] }
```

Comment:
Putting that sort before the group by, is this implicitly relying on ruby's preservation of hash insertion order? I feel like we could make a stronger guarantee if we sorted after grouping?
```suggestion
    group_by { |entry| [entry.category, entry.change] }.
    map { |key, items| [key, items.sort_by(&:subcategory)] }
```

Reply:
> Putting that sort before the group by, is this implicitly relying on ruby's preservation of hash insertion order? I feel like we could make a stronger guarantee if we sorted after grouping?

Good point! I'd wonder if it's much of a concern if the language itself makes those guarantees that hash order is preserved by insertion, but I'd be fine to make it more explicit, especially since it's not much to implement.

Reply:
> especially since it's not much to implement.

I spoke too soon. The grouping is not actually per top-level category, it's also baking in the "same change" grouping which combines separate pull requests with the same message.

```
{["Internal", "Improve changelog tool sorting and capitalization"]=>[#<struct ChangelogEntry>]
```

We might be able to change this to multiple passes (i.e. group by category, then sort, then reintroduce same-change grouping).

Reply:
I think I'll keep this as-is for now, unless you can think of an obvious workaround. The specs should help keep us honest, at least.

Reply:
that seems fine to me!